mds: scrub repair does not clear earlier damage health status

neesingh-rh opened the pull request:
Fixes: https://tracker.ceph.com/issues/54557
Signed-off-by: Neeraj Pratap Singh <neesingh@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

dparmar18 commented on qa/tasks/cephfs/test_scrub.py:
```diff
@@ -176,3 +177,40 @@ def test_scrub_backtrace(self):
+                         [a['dname'] for a in backtrace['ancestors']])
+
+        # Go corrupt the backtrace
+        self.fs._write_data_xattr(file_ino, "parent",
```
I think you should record and assert the damage count before this line

neesingh-rh replied:
Thanks, I did by mistake.

dparmar18 commented on src/mds/CInode.cc:
```diff
@@ -4780,6 +4780,8 @@ void CInode::validate_disk_state(CInode::validated_data *results,
                            false);
         // Flag that we repaired this BT so that it won't go into damagetable
         results->backtrace.repaired = true;
+        damage_entry_id_t damage_id = mdcache->mds->damage_table.inode_to_id(in->ino());
```
better name instead of `damage_id`? Maybe `damaged_inode_id`

neesingh-rh replied:
Replaced with `damaged_entry_id`

vshankar replied:
I don't fret about variable names, but, damage_id looks fine to me.

dparmar18 commented:
@neesingh-rh [L4781](https://github.com/ceph/ceph/blob/main/src/mds/CInode.cc#L4781) says that setting the `repaired` flag to true would prevent it's entry in damagetable. Seems like someone did address this issue in past but inode still sneaked into the damagetable?

neesingh-rh commented:
> @neesingh-rh [L4781](https://github.com/ceph/ceph/blob/main/src/mds/CInode.cc#L4781) says that setting the `repaired` flag to true would prevent it's entry in damagetable. Seems like someone did address this issue in past but inode still sneaked into the damagetable?

@dparmar18 Setting the `repaired` flag to true makes the ScrubStack.cc to realise that backtrace is repaired and hence, logs shows `Inode Repaired` but its not clearing the entry (earlier) from the damage table. That's why we have to manually erase it by `damage rm` command which shouldn't be.

dparmar18 commented:
> > @neesingh-rh [L4781](https://github.com/ceph/ceph/blob/main/src/mds/CInode.cc#L4781) says that setting the `repaired` flag to true would prevent it's entry in damagetable. Seems like someone did address this issue in past but inode still sneaked into the damagetable?
> 
> @dparmar18 Setting the `repaired` flag to true makes the ScrubStack.cc to realise that backtrace is repaired and hence, logs shows `Inode Repaired` but its not clearing the entry (earlier) from the damage table. That's why we have to manually erase it by `damage rm` command which shouldn't be. And that's issue which this PR fixes.

Ah, that comment is a bit misleading. Thanks for the explanation!

dparmar18 commented:
I'm fine with the current patch too but I was just thinking we can simplify this:

DamageTable
```
damage_entry_id_t DamageTable::remove_damaged_entry(inodeno_t ino) {
	erase(remotes.find(ino)->second->id);
}
```
---
CInode

```
mdcache->mds->damage_table.remove_damaged_entry(in->ino());
```

What do you think? @neesingh-rh

neesingh-rh commented:
> I'm fine with the current patch too but I was just thinking we can simplify this:
> 
> DamageTable
> 
> ```
> damage_entry_id_t DamageTable::remove_damaged_entry(inodeno_t ino) {
> 	erase(remotes.find(ino)->second->id);
> }
> ```
> 
> CInode
> 
> ```
> mdcache->mds->damage_table.remove_damaged_entry(in->ino());
> ```
> 
> What do you think? @neesingh-rh

@dparmar18 Yeah, it looks simpler. And I had thought about this but just to avoid confusion in `CInode.cc` between two function for erase(kind of) , I went with this. If u say second is better, will update. What say?

vshankar commented on src/mds/DamageTable.cc:
```diff
@@ -200,6 +200,11 @@ bool DamageTable::notify_remote_damaged(inodeno_t ino, std::string_view path)
   return false;
 }
 
+damage_entry_id_t DamageTable::inode_to_id(inodeno_t ino)
```
Does this handle cases when the damage (type) is a dirfrag or a dentry?

dparmar18 replied:
It does, https://github.com/ceph/ceph/blob/main/src/mds/DamageTable.cc#L259-L264

vshankar replied:
Can you explain how? `remotes` track damaged backtraces.

neesingh-rh replied:
> Can you explain how? `remotes` track damaged backtraces.

@vshankar It's not actually handling dirfrag, dentry. I guess @dparmar misunderstood. I am working on it.

dparmar18 replied:
my bad `inode_to_id` doesnt handle all cases, misunderstood it this fn with `erase()`

dparmar18 commented:
> > I'm fine with the current patch too but I was just thinking we can simplify this:
> > DamageTable
> > ```
> > damage_entry_id_t DamageTable::remove_damaged_entry(inodeno_t ino) {
> > 	erase(remotes.find(ino)->second->id);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > CInode
> > ```
> > mdcache->mds->damage_table.remove_damaged_entry(in->ino());
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What do you think? @neesingh-rh
> 
> @dparmar18 Yeah, it looks simpler. And I had thought about this but just to avoid confusion in `CInode.cc` between two function for erase(kind of) , I went with this.(Doesn't have any strong opinion) If u say second is better, will update. What say?

@neesingh-rh This would introduce one more `erase()` call in `CInode.cc` right? Doesn't that lead to more confusion? Or I didn't understand correctly?

mchangir commented on src/mds/DamageTable.cc:
```diff
@@ -200,6 +201,29 @@ bool DamageTable::notify_remote_damaged(inodeno_t ino, std::string_view path)
+
+void DamageTable::remove_dirfrag_damage_entry(CDir *diri)
+{
+  DamageTable::erase(dirfrags.find(DirFragIdent(diri->inode->ino(), diri->frag))->second->id);
```
what if the directory inode hasn't been added to the `dirfrags` in DamageTable ?

also, the argument name should just be `dir` and not `diri` ... since `diri` implies the directory inode and not a directory

neesingh-rh replied:
> what if the directory inode hasn't been added to the `dirfrags` in DamageTable ?
There is check for the `damage_id` in the `DamageTable::erase()` but to ensure and reduce the iteration prior to that in case of its absence, i will put a check before that.

mchangir commented on src/mds/DamageTable.h:
```diff
@@ -155,6 +156,12 @@ class DamageTable
 
+    void remove_dentry_damage_entry(CDir *dir);
+
+    void remove_dirfrag_damage_entry(CDir *diri);
```
nit: typo
rename `diri` to `dir`

neesingh-rh replied:
ACK

mchangir commented on src/mds/DamageTable.cc:
```diff
@@ -200,6 +201,29 @@ bool DamageTable::notify_remote_damaged(inodeno_t ino, std::string_view path)
+    for (const auto &i : by_id){
+      for (const auto &dn : frag_dentries){
+        if (i.second == dn.second){
+          DamageTable::erase(i.first);
```
I was concerned for the matching algorithm time complexity being O(m x n) and was wondering if it can be optimized.

But maybe there won't be too many damaged entries in the table for this issue to be of real concern.

neesingh-rh replied:
Yup, I too was concerned about the complexity but couldn't think of that's why choose this way unlike others damage types.Will re-think.

kotreshhr commented:
@neesingh-rh No test for this ?

neesingh-rh commented:
> @neesingh-rh No test for this ?

Will update with the tests soon.

dparmar18 commented on src/mds/DamageTable.cc:
```diff
@@ -200,6 +201,37 @@ bool DamageTable::notify_remote_damaged(inodeno_t ino, std::string_view path)
 
+void DamageTable::remove_dentry_damage_entry(CDir *dir)
+{
+  if (dentries.count(
+        DirFragIdent(dir->inode->ino(), dir->frag)
+        ) > 0){
+          const std::map<DentryIdent, DamageEntryRef> &frag_dentries =
+            dentries.at(DirFragIdent(dir->inode->ino(), dir->frag));
+          for (const auto &dn : frag_dentries){
+            for (const auto &i : by_id){
+              if (i.second == dn.second){
+                DamageTable::erase(i.first);
+              }
+            }
+          }
+        }
```
```
set<DamageEntryRef> dmg_entries:

for(const auto &f_g : frag_dentries) {
	dmg_entries.insert(f_g->second);
}

for(const auto &i : by_id) {
	if(i->second != dmg_entries.end()) {
		DamageTable::erase(i.first);
	}
}
```

this should reduce the TC. Makes sense?

neesingh-rh commented on src/mds/DamageTable.cc:
```diff
@@ -200,6 +201,37 @@ bool DamageTable::notify_remote_damaged(inodeno_t ino, std::string_view path)
 
+void DamageTable::remove_dentry_damage_entry(CDir *dir)
+{
+  if (dentries.count(
```
`is_dentry_damaged` needs `snap_id` and `dname`and we just need to check if entry is there or not which can be only done by `dentries.count()` and anyways erase deletes the complete `DirFragIdent(ino,frag)` entry for the damage id.That's why i thought its better to have  a check for `dentries.count()` here .

vshankar commented on src/mds/DamageTable.cc:
```diff
@@ -200,6 +201,37 @@ bool DamageTable::notify_remote_damaged(inodeno_t ino, std::string_view path)
+          for (const auto &dn : frag_dentries){
+            for (const auto &i : by_id){
+              if (i.second == dn.second){
+                DamageTable::erase(i.first);
```
Basically, this is trying to find all the damage ids for the CDir, isn't it? You could just iterate `std::map<DentryIdent, DamageEntryRef>` for the `DirFragIdent`. Wouldn't that suffice?

neesingh-rh replied:
Yeah, we are anyways deleting the `DirFragIdent(ino,frag)` entry. We can directly iterate over it. Thanks

neesingh-rh commented on src/mds/DamageTable.cc:
```diff
@@ -200,6 +201,37 @@ bool DamageTable::notify_remote_damaged(inodeno_t ino, std::string_view path)
+void DamageTable::remove_dirfrag_damage_entry(CDir *dir)
+{
+  if (is_dirfrag_damaged(dir)){
+    DamageTable::erase(dirfrags.find(DirFragIdent(dir->inode->ino(), dir->frag))->second->id);
```
ACK

vshankar commented:
@neesingh-rh (re: test case failure). Where does the test (https://github.com/ceph/ceph/pull/48895/files#diff-33da6740c9b71b9d4171912814facb2425ea98a800b1a6040b4ebf769cb35c36R384) fail exactly?

neesingh-rh commented:
> @neesingh-rh (re: test case failure). Where does the test (https://github.com/ceph/ceph/pull/48895/files#diff-33da6740c9b71b9d4171912814facb2425ea98a800b1a6040b4ebf769cb35c36R384) fail exactly?

At this point, https://github.com/ceph/ceph/pull/48895/files#diff-33da6740c9b71b9d4171912814facb2425ea98a800b1a6040b4ebf769cb35c36R427 where it is not removing the `backtrace` damage created for path `testdir/hardlink`.

vshankar commented:
> > @neesingh-rh (re: test case failure). Where does the test (https://github.com/ceph/ceph/pull/48895/files#diff-33da6740c9b71b9d4171912814facb2425ea98a800b1a6040b4ebf769cb35c36R384) fail exactly?
> 
> At this point, https://github.com/ceph/ceph/pull/48895/files#diff-33da6740c9b71b9d4171912814facb2425ea98a800b1a6040b4ebf769cb35c36R427 where it is not removing the `backtrace` damage created for path `testdir/hardlink`.

You removed dirfrag. You'd need to run cephfs-data-scan (scan_extents,scan_inodes,scan_links) to regenerate the directory object from the data pool objects.

vshankar commented:
jenkins retest this please

vshankar commented on src/mds/DamageTable.cc:
```diff
@@ -200,6 +201,33 @@ bool DamageTable::notify_remote_damaged(inodeno_t ino, std::string_view path)
+  if (dentries.count(
+        DirFragIdent(dir->inode->ino(), dir->frag)
+        ) > 0){
+          const std::map<DentryIdent, DamageEntryRef> &frag_dentries =
```
nit: use `auto`

neesingh-rh replied:
ACK

vshankar commented on src/mds/DamageTable.cc:
```diff
@@ -200,6 +201,33 @@ bool DamageTable::notify_remote_damaged(inodeno_t ino, std::string_view path)
+          const std::map<DentryIdent, DamageEntryRef> &frag_dentries =
+            dentries.at(DirFragIdent(dir->inode->ino(), dir->frag));
+          for(const auto &i : frag_dentries) {
+            erase(i.second->id);
```
This _could_ remove an entry from the container that's being iterated.

neesingh-rh replied:
ACK

vshankar commented:
jenkins retest this please

neesingh-rh commented:
jenkins test api

vshankar reviewed (commented):
Looks fine. Will include this in my next integration run.

vshankar commented on qa/tasks/cephfs/test_forward_scrub.py:
```diff
@@ -298,3 +299,208 @@ def test_backtrace_repair(self):
+            self.fs.mon_manager.raw_cluster_cmd(
+                'tell', 'mds.{0}'.format(self.fs.get_active_names()[0]),
+                "damage", "ls", '--format=json-pretty'))
+        self.assertEqual(len(damage), 1)
```
@neesingh-rh This check fails with run: https://pulpito.ceph.com/vshankar-2023-03-02_05:17:07-fs-wip-vshankar-testing-20230228.105516-testing-default-smithi/7191542/

`dirfrag ls` gives:
```
2023-03-02T06:03:40.840 INFO:teuthology.orchestra.run.smithi119.stdout:[
2023-03-02T06:03:40.840 INFO:teuthology.orchestra.run.smithi119.stdout:    {
2023-03-02T06:03:40.840 INFO:teuthology.orchestra.run.smithi119.stdout:        "damage_type": "backtrace",
2023-03-02T06:03:40.841 INFO:teuthology.orchestra.run.smithi119.stdout:        "id": 561354219,
2023-03-02T06:03:40.841 INFO:teuthology.orchestra.run.smithi119.stdout:        "ino": 1099511627777,
2023-03-02T06:03:40.841 INFO:teuthology.orchestra.run.smithi119.stdout:        "path": "/testdir/hardlink"
2023-03-02T06:03:40.841 INFO:teuthology.orchestra.run.smithi119.stdout:    },
2023-03-02T06:03:40.842 INFO:teuthology.orchestra.run.smithi119.stdout:    {
2023-03-02T06:03:40.842 INFO:teuthology.orchestra.run.smithi119.stdout:        "damage_type": "dir_frag",
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "id": 2832515251,
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "ino": 1099511627776,
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "frag": "*",
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "path": "/dir"
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:    },
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:    {
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "damage_type": "backtrace",
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:        "id": 3143126501,
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:        "ino": 1099511627776,
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:        "path": "/dir"
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:    }
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:]
```

PTAL

neesingh-rh replied:
@vshankar During debugging for above failure, i.e. the `test_health_status_after_dirfrag_repair` I manually added more inode damages in this particular test and seeing that only the `dir_frag` damage_entry for the latest/last inode damage is being removed from the damage table. But on checking the mds logs, it shows that the repair call in `CInode.cc` for each and every inode is being visited.

neesingh-rh replied:
> @vshankar During debugging for above failure, i.e. the `test_health_status_after_dirfrag_repair` I manually added more inode damages in this particular test and seeing that only the `dir_frag` damage_entry for the latest/last inode damage is being removed from the damage table. But on checking the mds logs, it shows that the repair call in `CInode.cc` for each and every inode is being visited.

I see that it fails at the check for `is_dirfrag_damaged()` before the call of `erase()`

neesingh-rh replied:
@vshankar Here's the run: https://pulpito.ceph.com/dparmar-2023-03-13_11:40:03-fs:functional-wip-vshankar-testing-20230228.105516-distro-default-smithi/

vshankar replied:
@neesingh-rh checking back - you fixed the original issue I reported but now you are running into a different one?

vshankar replied:
@neesingh-rh This is a new failure where the test fails expecting a cluster log with message "bad backtrace on inode". The second time scrub is run, it does not traverse `/dir_test/file`:

```
2023-03-13T12:11:32.252+0000 7fd00043e700 10  mds.0.cache.snaprealm(0x1 seq 1 0x562638154a00) get_snaps  (seq 1 cached_seq 1)
2023-03-13T12:11:32.252+0000 7fd00043e700 15 mds.0.scrubstack scrub_dirfrag skip dentry (file,head), no change since last scrub
2023-03-13T12:11:32.252+0000 7fd00043e700 20 mds.0.cache.dir(0x10000000000) get_num_head_items() = 1; fnode.fragstat.nfiles=1 fnode.fragstat.nsubdirs=0
```

Thus the backtrace cluster log warning is not generated.

vshankar commented:
@neesingh-rh I couldn't look into this last week. Let me know if you made progress with debugging the failure, else, I'll surely have a look this week.

neesingh-rh commented:
> @neesingh-rh I couldn't look into this last week. Let me know if you made progress with debugging the failure, else, I'll surely have a look this week.

I couldn't make any much progress into this.

vshankar commented:
@neesingh-rh On this today. Will post an update.

vshankar commented:
@neesingh-rh https://github.com/ceph/ceph/pull/48895#discussion_r1146496730

vshankar commented:
marking ready for review as its close to gettin done.

neesingh-rh commented:
https://pulpito.ceph.com/neesingh-2023-05-18_10:51:51-fs-wip-neesingh-testing-54557-distro-default-smithi/

vshankar commented:
> https://pulpito.ceph.com/neesingh-2023-05-18_10:51:51-fs-wip-neesingh-testing-54557-distro-default-smithi/

Will check. thx!

vshankar commented:
> > https://pulpito.ceph.com/neesingh-2023-05-18_10:51:51-fs-wip-neesingh-testing-54557-distro-default-smithi/
> 
> Will check. thx!

Checking now. Will update.

neesingh-rh commented:
jenkins retest this please

neesingh-rh commented:
I did run the teuthology after cdoing cleanups and here's the link:https://pulpito.ceph.com/neesingh-2023-11-23_10:04:20-fs-wip-neesingh-testing-231123-distro-default-smithi/

neesingh-rh commented:
@vshankar PTAL

vshankar commented:
> I did run the teuthology after cdoing cleanups and here's the link:https://pulpito.ceph.com/neesingh-2023-11-23_10:04:20-fs-wip-neesingh-testing-231123-distro-default-smithi/

Could you explain what the changes are that fixes that test case failure reported in https://github.com/ceph/ceph/pull/48895#pullrequestreview-1321439000 ?

neesingh-rh commented:
> > I did run the teuthology after cdoing cleanups and here's the link:https://pulpito.ceph.com/neesingh-2023-11-23_10:04:20-fs-wip-neesingh-testing-231123-distro-default-smithi/
> 
> Could you explain what the changes are that fixes that test case failure reported in [#48895 (review)](https://github.com/ceph/ceph/pull/48895#pullrequestreview-1321439000) ?

As we discussed earlier that after debuging the code many times it seemed that there was no problem in the code, we need to check only for the test case.
There were two failures that were happening:
1. `assert` was failing on not finding `MDS_DAMAGE` in the `get_mon_health()` dict which got solved by adding sleep of some time cause addition of `MDS_DAMAGE` in the dict takes some time.
2. It was skipping the scrubbing second time when `scrub start repair` was run which is only responsible for removing the damage from damage list. When I looked the code in `ScrubStack.cc`, It skips the scrubbing when there's no change since last scrub unless the header contains `force` flag and in our case we need to run `scrub start repair` second time , so I added force in the `scrub start repair` for `test_health_status_after_backtrace_repair`

dparmar18 commented:
> > > I did run the teuthology after cdoing cleanups and here's the link:https://pulpito.ceph.com/neesingh-2023-11-23_10:04:20-fs-wip-neesingh-testing-231123-distro-default-smithi/
> > 
> > 
> > Could you explain what the changes are that fixes that test case failure reported in [#48895 (review)](https://github.com/ceph/ceph/pull/48895#pullrequestreview-1321439000) ?
> 
> As we discussed earlier that after debuging the code many times it seemed that there was no problem in the code, we need to check only for the test case. There were two failures that were happening:
> 
> 1. `assert` was failing on not finding `MDS_DAMAGE` in the `get_mon_health()` dict which got solved by adding sleep of some time cause addition of `MDS_DAMAGE` in the dict takes some time.

maybe you can use `wait_until_true()` with a timeout? 
> 2. It was skipping the scrubbing second time when `scrub start repair` was run which is only responsible for removing the damage from damage list. When I looked the code in `ScrubStack.cc`, It skips the scrubbing when there's no change since last scrub unless the header contains `force` flag and in our case we need to run `scrub start repair` second time , so I added force in the `scrub start repair` for `test_health_status_after_backtrace_repair`

vshankar commented:
> > > > > I did run the teuthology after cdoing cleanups and here's the link:https://pulpito.ceph.com/neesingh-2023-11-23_10:04:20-fs-wip-neesingh-testing-231123-distro-default-smithi/
> > > > 
> > > > 
> > > > Could you explain what the changes are that fixes that test case failure reported in [#48895 (review)](https://github.com/ceph/ceph/pull/48895#pullrequestreview-1321439000) ?
> > > 
> > > 
> > > As we discussed earlier that after debuging the code many times it seemed that there was no problem in the code, we need to check only for the test case. There were two failures that were happening:
> > > 
> > > 1. `assert` was failing on not finding `MDS_DAMAGE` in the `get_mon_health()` dict which got solved by adding sleep of some time cause addition of `MDS_DAMAGE` in the dict takes some time.
> > 
> > 
> > maybe you can use `wait_until_true()` with a timeout?
> 
> We can but if there's no harm lets stick to this

I'm with @dparmar18 on this one. Using wait_until_true() is the preferred way rather than using sleep.

vshankar commented:
Otherwise LGTM.

neesingh-rh commented:
Wait, its failing after the latest changes.

vshankar commented:
jenkins test windows

vshankar commented on qa/tasks/cephfs/test_forward_scrub.py:
```diff
@@ -46,6 +47,12 @@ def _get_paths_to_ino(self):
         return inos
 
+    def _check_for_presence_of_MDS_damage(self):
+        return self.assertIn("MDS_DAMAGE", self.mds_cluster.mon_manager.get_mon_health()['checks'])
```
return an assert?

This should be

```
return "MDS_DAMAGE" in self.mds_cluster.mon_manager.get_mon_health()['checks']
```

neesingh-rh replied:
I thought return value should be a boolean and assert also returns a boolean. that's why used it.

vshankar commented on qa/tasks/cephfs/test_forward_scrub.py:
```diff
@@ -46,6 +47,12 @@ def _get_paths_to_ino(self):
+        return self.assertIn("MDS_DAMAGE", self.mds_cluster.mon_manager.get_mon_health()['checks'])
+
+    def _check_for_absence_of_MDS_damage(self):
+        return self.assertNotIn("MDS_DAMAGE", self.mds_cluster.mon_manager.get_mon_health()['checks'])
```
ditto.

neesingh-rh commented:
Run link: https://pulpito.ceph.com/neesingh-2023-11-29_05:06:25-fs-wip-neesingh-testing-231123-distro-default-smithi/

dparmar18 commented on qa/tasks/cephfs/test_forward_scrub.py:
```diff
@@ -46,6 +47,12 @@ def _get_paths_to_ino(self):
 
         return inos
 
+    def _check_for_presence_of_MDS_damage(self):
+        return "MDS_DAMAGE" in self.mds_cluster.mon_manager.get_mon_health()['checks']
+
+    def _check_for_absence_of_MDS_damage(self):
+        return "MDS_DAMAGE" not in self.mds_cluster.mon_manager.get_mon_health()['checks']
```
do we really need two functions for doing almost the same thing? how about having just method named `_is_mds_damage()` (or any other similar name) and negate it while using when you want no mds damage

neesingh-rh replied:
Updated, PTAL

dparmar18 reviewed (approved):
LGTM

dparmar18 commented:
jenkins test make check arm64

vshankar commented:
fs+smoke suite - https://pulpito.ceph.com/?branch=wip-vshankar-testing-20231206.125818

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/Main#06-Dec-2023
